new_ai.py

```diff
@@ -0,0 +1,232 @@
+import heapq
+from typing import List, Optional, Tuple
+from dataclasses import dataclass
+from collections import deque
+from utils import Coordinate, Direction, GameState, Color, MirrorAngle, Shape
+
+@dataclass(frozen=True)
+class ResolutionState:
+    pawns: List[Coordinate]# Positions of all pawns
+    cost: int 
+    heuristic: int  # A* heuristic value
+    previous_state: Optional["ResolutionState"] = None
+
+    # Comparison function for the priority queue based on cost + heuristic
+    def __lt__(self, other: "ResolutionState") -> bool:
+        return (self.cost + self.heuristic) < (other.cost + other.heuristic)
+
+    # Generates the sequence of moves from the current state to the initial state
+    def get_move_sequence(self) -> List[Tuple[Color, Coordinate]]:
+        moves = []
+        current = self
+     # Traverse back through previous states to generate the sequence of moves
+        while current.previous_state is not None:
+            for pawn_color, (curr_pos, prev_pos) in enumerate(
+                zip(current.pawns, current.previous_state.pawns)
+            ):
+                if curr_pos != prev_pos:
+                    moves.append((Color(pawn_color), curr_pos))# Record the move of the pawn
+                    break
+            current = current.previous_state
+        return list(reversed(moves)) # Return the sequence in the correct order
+
+    # Custom string representation for debugging and output
+    def __str__(self):
+        pawn_moved = None
+
+        if self.previous_state:
+          # Find the pawn that was moved in this state
+            for i, (curr_pos, prev_pos) in enumerate(
+                zip(self.pawns, self.previous_state.pawns)
+            ):
+                if curr_pos != prev_pos:
+                    pawn_moved = i
+                    break
+
+        string = "\n\n"
+
+        if pawn_moved is None:
+            string += "Initial state:\n"
+        else:
+            string += f"Move pawn {get_color_name(Color(pawn_moved))} with a cost of {self.cost}:\n"
+        string += "\n".join(
+            f"  Pawn {i}: ({pos.x}, {pos.y})" + (" <-" if i == pawn_moved else "")
+            for i, pos in enumerate(self.pawns)
+        )
+
+        return string
+    
+# Helper function to convert color code to its name
+def get_color_name(color_code: Color):
+    colors = ["red", "green", "blue", "yellow"]
+    return colors[color_code.value]
+
+def get_shape(shape_code: Shape):
+    shapes = ["circle", "square", "triangle", "star"]
+    return shapes[shape_code.value]
+
+# AI player class that will use A* to find the solution
+class AIPlayer:

```

Comment:
```suggestion
class AIPlayer(GameResolutionInterface):
```

You forgot to make your class implement the `GameResolutionInterface`. You might have to change the `solve` method to `resolve`

Reply:
Sorry, I'll change it .

Reply:
No problem haha